File: rails/learning_rails/app/channels/room_channel.rb
class RoomChannel < ApplicationCable::Channel
  # 購読 => チャネルとクライアントとの関連付けのこと
  # 購読後に呼ばれる
  def subscribed
    # ブロードキャスト用のストリーム名を設定
    stream_from "room_channel"
  end

  # 購読解除後に呼ばれる
  def unsubscribed
  end

  # speakメソッド
  def speak(data)
    # テキストを保存
    message = Message.create!(content: data["message"])
    ActionCable.server.broadcast(
      # ③ストリーム名"room_channel"にmessageを送信
      "room_channel", { message: render_message(message) }
    )
  end

  private

  def render_message(message)
    # テンプレートのレンダリングを行う
    ApplicationController.render(
      partial: "messages/message",
      locals: { message: message }
    )
  end
end

File: rails/learning_rails/app/javascript/channels/room_channel.js
import consumer from "./consumer"

consumer.subscriptions.create("RoomChannel", {
  // 接続時
  connected() {
    document.
      querySelector('input[data-behavior="room_speaker"]').
      addEventListener('keypress', (event) => {
        if (event.key === 'Enter') {
          // ①エンターキーが押された時にspeakメソッドを呼び出す
          this.speak(event.target.value);
          // フォームを初期化
          event.target.value = '';
          // イベントを終了
          return event.preventDefault();
        }
      });
  },

  // 切断時
  disconnected() {
  },

  // ④サーバーからの通信を受信したとき
  received(data) {
    const element = document.querySelector('#messages')
    // テキストを最後の子要素の後に挿入する
    element.insertAdjacentHTML('beforeend', data['message'])
  },

  // speakメソッド
  speak: function (message) {
    // ②サーバーサイドのspeakメソッドを呼び出す
    return this.perform('speak', {message: message});
  },
});
